src/sentry/taskworker/task.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import timedelta
+from functools import update_wrapper
+from typing import TYPE_CHECKING, Generic, ParamSpec, TypeVar
+from uuid import uuid4
+
+import orjson
+from django.conf import settings
+from django.utils import timezone
+from google.protobuf.timestamp_pb2 import Timestamp
+from sentry_protos.sentry.v1.taskworker_pb2 import RetryState, TaskActivation
+
+from sentry.taskworker.retry import Retry
+
+if TYPE_CHECKING:
+    from sentry.taskworker.registry import TaskNamespace
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+class Task(Generic[P, R]):
+    def __init__(
+        self,
+        name: str,
+        func: Callable[P, R],
+        namespace: TaskNamespace,
+        retry: Retry | None,
+        idempotent: bool = False,
+        deadline: timedelta | int | None = None,
+    ):
+        self.name = name
+        self._func = func
+        self._namespace = namespace
+        self._retry = retry
+        self._idempotent = idempotent
+        self._deadline = deadline
+        update_wrapper(self, func)
+
+    @property
+    def retry(self) -> Retry | None:
+        return self._retry
+
+    @property
+    def idempotent(self) -> bool:
+        return self._idempotent or False
+
+    @property
+    def deadline_timestamp(self) -> int | None:
+        # TODO add namespace/default deadlines
+        if not self._deadline:
+            return None
+        if isinstance(self._deadline, int):
+            return int(timezone.now().timestamp() + self._deadline)
+        if isinstance(self._deadline, timedelta):
+            return int(timezone.now().timestamp() + self._deadline.total_seconds())
+        raise ValueError(f"unknown type for Task.deadline {self._deadline}")
```

Comment:
How do we intend to use this? Since we calculate the deadline from now, it's always going to be increasing. I feel like this could be a footgun, where someone rechecks this on the Task and gets a new deadline each time.

Reply:
Deadlines haven't been implemented yet. I'll remove all of this and we can revisit this when we have a bit more substance to build on.